flink-tests/src/test/java/org/apache/flink/test/execution/JobStatusChangedListenerITCase.java

```diff
@@ -0,0 +1,113 @@
+import org.apache.flink.core.execution.DefaultJobExecutionStatusEvent;
+import org.apache.flink.core.execution.JobExecutionStatusEvent;
+import org.apache.flink.core.execution.JobStatusChangedEvent;
+import org.apache.flink.core.execution.JobStatusChangedListener;
+import org.apache.flink.core.execution.JobStatusChangedListenerFactory;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.runtime.execution.DefaultJobCreatedEvent;
+import org.apache.flink.util.CloseableIterator;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.configuration.DeploymentOptions.JOB_STATUS_CHANGED_LISTENERS;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for job status changed listener. */
+public class JobStatusChangedListenerITCase {
+    private static List<JobStatusChangedEvent> statusChangedEvents = new ArrayList<>();
+
+    @Test
+    void testJobStatusChanged() throws Exception {
+        Configuration configuration = new Configuration();
+        configuration.set(
+                JOB_STATUS_CHANGED_LISTENERS,
+                Collections.singletonList(TestingJobStatusChangedListenerFactory.class.getName()));
+        try (StreamExecutionEnvironment env =
+                StreamExecutionEnvironment.getExecutionEnvironment(configuration)) {
+            List<String> sourceValues = Arrays.asList("a", "b", "c");
+            List<String> resultValues = new ArrayList<>();
+            try (CloseableIterator<String> iterator =
+                    env.fromCollection(sourceValues).executeAndCollect()) {
+                while (iterator.hasNext()) {
+                    resultValues.add(iterator.next());
+                }
+            }
+            assertThat(resultValues).containsExactlyInAnyOrder(sourceValues.toArray(new String[0]));
+        }
+        assertThat(statusChangedEvents.size()).isEqualTo(3);
+        assertThat(statusChangedEvents.get(0).jobId())
+                .isEqualTo(statusChangedEvents.get(1).jobId());
+        assertThat(statusChangedEvents.get(0).jobName())
+                .isEqualTo(statusChangedEvents.get(1).jobName());
+
+        assertThat(statusChangedEvents.get(1).jobId())
+                .isEqualTo(statusChangedEvents.get(2).jobId());
+        assertThat(statusChangedEvents.get(1).jobName())
+                .isEqualTo(statusChangedEvents.get(2).jobName());
+
+        statusChangedEvents.forEach(
+                event -> {
+                    if (event instanceof DefaultJobExecutionStatusEvent) {
+                        JobExecutionStatusEvent status = (JobExecutionStatusEvent) event;
+                        assertThat(
+                                        (status.oldStatus() == JobStatus.CREATED
+                                                        && status.newStatus() == JobStatus.RUNNING)
```

Comment:
can we add a test that check for a :`FAILING` / `FAILED` and for `CANCELLING`:`CANCELED`:

Reply:
Good suggestion. Added test cases accordingly. It actually help me to find out that the job status changed listener should be added for MiniClusterExecutor.